Please fill out the information of your group!

| <p style="text-align: center;">First Name</p>  | <p style="text-align: center;">Family Name</p> | <p style="text-align: left">Matr.-No.</p> |
| ---------------------------------------------- | ---------------------------------------------- | -------- |
| <p style="text-align: left">*Cesar Miguel*</p>| <p style="text-align: left">*Valdez Cordova*</p> | <p style="text-align: left">*K12046803*</p>  |
| <p style="text-align: left">*Ionelia*</p>| <p style="text-align: left">*Buzatu*</p> | <p style="text-align: left">*k12008243*</p>  |


<h2 style="text-align: center">344.063: Special Topics - Natural Language Processing with Deep Learning (SS2022)</h2>
<h1 style="color:rgb(0,120,170)">Assignment 1: Open</h2>

<div style="background-color:rgb(224, 243, 255)">
<b>Terms of Use</b><br>
This  material is prepared for educational purposes at the Johannes Kepler University (JKU) Linz, and is exclusively provided to the registered students of the mentioned course at JKU. It is strictly forbidden to distribute the current file, the contents of the assignment, and its solution. The use or reproduction of this manuscript is only allowed for educational purposes in non-profit organizations, while in this case, the explicit prior acceptance of the author(s) is required.

**Author:** Navid Rekab-saz<br>
**Email:** navid.rekabsaz@jku.at<br>

</div>

<h2>Table of contents</h2>
<ol>
    <a href="#section-general-guidelines"><li style="font-size:large;font-weight:bold">General Guidelines</li></a>
    <a href="#section-dummy"><li style="font-size:large;font-weight:bold">Task A: Collection Processing and Dummy Baseline (10 points)</li></a>
    <a href="#section-model1"><li style="font-size:large;font-weight:bold">Task B: Implementation of Model I (15 points)</li></a>
    
</ol>

<a name="section-general-guidelines"></a><h2 style="color:rgb(0,120,170)">General Guidelines</h2>

<div style="background-color:rgb(224, 243, 255)">

### Assignment objective
This Notebook provide a general instruction for the Open assignment. This Notebook should encompas all aspects of the assignment such as the implemented codes and reports. The cells can contains code, reports, charts, tables, or any other material, required for the assignment. Cover the questions/points, mentioned in the tasks, but also add any necessary point for understanding your experiments. Try to provide the solutions in a clear, and visual way! 

Please discuss any unclear point in the assignment with the author(s) of the assignment.

</div>

<div style="background-color:rgb(224, 243, 255)">

### Selected NLP Task and Test Collection 

In this assignment, you can approach any natural language processing task, such as text summarization, question answering, machine translation, etc. To do the assignment, you should first select an **available standard test collection** for the task of your choice. A standard test collection means a dataset that provides training/validation/test sets, and has clearly defined methods/metrics for performance evaluation. Examples of such collections are the ones in various ML challenges (e.g. in Kaggle).

If you have any doubts or questions about the selected collection, feel free to resolve it with the lecturer(s) and tutor(s): <navid.rekabsaz@jku.at>, <shahed.masoudian@jku.at>
</div>

<div style="background-color:rgb(224, 243, 255)">

### Libraries

The assignment should be implemented with recent versions of `Python` (>3.7) and `PyTorch` (>1.7). Any standard Python library can be used, so far that the library is free and can be simply installed using `pip` or `conda`. Examples of potentially useful libraries are `transformer`, `scikit-learn`, `numpy`, `scipy`, `gensim`, `nltk`, `spaCy`, and `AllenNLP`. Use the latest stable version of each library.

</div>


<div style="background-color:rgb(224, 243, 255)">

### Submission

Each group should submit the following two files:

- One Jupyter Notebook file (`.ipynb`), containing all the code, results, visualizations, etc. **In the submitted Notebook, all the results and visualizations should already be present, and can be observed simply by loading the Notebook in a browser.** The Notebook must be self-contained, meaning that (if necessary) one can run all the cells from top to bottom without any error. Do not forget to put in your names and student numbers in the first cell of the Notebook. 
- The HTML file (`.html`) achieved from exporting the Jupyter Notebook to HTML (Download As HTML).

You do not need to include the data files in the submission.


</div>

<div style="background-color:rgb(224, 243, 255)">

### Publishing Experiments Results

It is encouraged that you log and store any information about the training and evaluation of the models in an ML dashboard like [`TensorBoard`](https://www.tensorflow.org/tensorboard) or [`wandb`](https://wandb.ai/site). This can contain any important aspect of training such as the changes in the evaluation results on validation, training loss, or learning rate. 

To this end, in the case of `TensorBoard`, after finalizing all experiments and cleaning any unnecessary experiment, publish the log files results through [`TensorBoard.dev`](https://tensorboard.dev). A simple way of doing it is by running the following command in the folder of log files:

`tensorboard dev upload --name my_exp --logdir path/to/output_dir`

`TensorBoard.dev` uploads the necessary files and provides a URL to see the TensorBoard's console. Insert the URL in the cell below.

</div>


**URL :** *EDIT!*

<a name="section-dummy"></a><h2 style="color:rgb(0,120,170)">Task A: Collection Processing and Dummy Baseline (10 points)</h2>

<div style="background-color:rgb(224, 243, 255)">

The objective of this task is to load and process the collection, to implement a simple dummy baseline, followed by its evaulation. Indeed, the actual functionality of each part depends on the selected collection. However, the following general steps are expected in the provided solution. 

**Loading the collection and preprocessing the data (2 point).** 

**Defining a dummy baseline and training it (2 point):** A dummy baseline can be any non-deep-learning model, which is considered as a weak baseline for further comparisons. This model can be a classical machine learning algorithm (like by using linear regression), or even the models that ignore input features such as the ones in [`DummyClassifier`](https://scikit-learn.org/stable/modules/generated/sklearn.dummy.DummyClassifier.html).

**Evaluation (2 point).**

**Overall functionality (2 point).**

**Reporting (2 point):** During loading and processing the collection, provide sufficient information and examples about the data and the applied processing steps. Report the evaluation results of the baseline model on the validation and test set in a table.

</div>

In [34]:
from huggingmolecules import MatModel, MatFeaturizer, MatConfig

import getpass
whoami = getpass.getuser()
print(f"user: {whoami}")

# The following import works only from the source code directory:
USERS_PATH_TO_REPO = {'semibah': '/home/semibah/huggingmolecules', 'ionelia': '/home/ionelia/pycharm-projects/huggingmolecules', 'other_user': 'please fill in here path'}
try:
    sys.path.append(USERS_PATH_TO_REPO[whoami])
except KeyError:
    raise("who are you? - add your username to the dict above.")
    
from experiments.src import TrainingModule, get_data_loaders
from experiments.src.training import training_metrics

import sys
import numpy as np
import pandas as pd
from rdkit import Chem
from rdkit.Chem import Descriptors, Lipinski
import torch
import torch.nn as nn
import torchmetrics
from torch.nn import MSELoss, CrossEntropyLoss, BCEWithLogitsLoss
from torch.optim import Adam
import torch.nn.utils.prune as prune

from torchinfo import summary
from tdc.utils import retrieve_label_name_list
from typing import Tuple, List, Optional, Any
import inspect

from pytorch_lightning import Trainer
from pytorch_lightning.metrics import MeanSquaredError, Metric
from pytorch_lightning.metrics.functional.classification import auroc

from tdc.single_pred import ADME
from sklearn.metrics import mean_absolute_error as mae
from sklearn.metrics import accuracy_score as acc


# the models used in task A
from sklearn.discriminant_analysis import LinearDiscriminantAnalysis
from sklearn.ensemble import ExtraTreesRegressor

Below we show **two problems of drug discovery** and apply in task A a linear regression for the **classification task** and tree regression model for the **regression task**. \
We will then implement a version of the recent Transformer algorithm adapting it to our drug discovery tasks. For each task there is a short description and an example of how it looks like. \
The loading of the data uses two packages: **tdc and huggingmolecules**, in **task A** for simplicity we use tdc and in **task B** because of the transformers we use: \
https://tdcommons.ai/single_pred_tasks/adme/#hydration-free-energy-freesolv but essentially they load the same data with the difference that tdc addapts it to the transformers architecture.

In [ ]:
# some functions we will need later

def drug_descriptors(drugs):
    """
    used to preprocess drug discovery data. this kind of data comes as a string format
    and here we extract information of each drug with the help of rdkit package. 
    each desciptor is an attribute of the drug e.g. the pH and, 
    all together, those attributes form the input for the model for a single sample.
    """
    
    def descriptors_array(mol):
        descriptors = np.array(get_descriptors.ComputeProperties(mol))
        return descriptors
    
    if isinstance(drugs, str):
        mol = Chem.MolFromSmiles(drug)
        drug_descriptors = []
        if mol:
            drug_descriptors = descriptors_array(mol) 
        return drug_descriptors
    else:
        dataset_descriptors = []
        for drug in drugs:
            mol = Chem.MolFromSmiles(drug)
            descriptors = descriptors_array(mol)
            if len(descriptors):
                dataset_descriptors.append(descriptors)
        return dataset_descriptors  
    
def regression_task(model, x_train, y_train, x_test, y_test):
    """
    train a regression model.
    return: the mean absolute error of the testset. 
    """
    model.fit(x_train, y_train)
    y_pred = model.predict(x_freesolv_test)
    mean_absolute_error = mae(y_test, y_pred)
    return mean_absolute_error

### Classification task

**HIA is a classification task and stands for human intestinal absorption.**

**Dataset Description**: When a drug is orally administered, it needs to be absorbed from the human gastrointestinal
system into the bloodstream of the human body. This ability of absorption is called human intestinal absorption
(HIA) and it is crucial for a drug to be delivered to the target.

**Task Description**: Binary classification. Given a drug SMILES string, predict the activity of HIA.

**Dataset Statistics**: 578 drugs.

**Suggested data split**: scaffold split.

**Evaluation Metric**: AUROC.

In [9]:
# loading classification data and split it into train, validation and test.
data_hia = ADME(name = 'HIA_Hou')
split_data_hia = data_hia.get_split(method='scaffold', frac=[0.7, 0.0, 0.3])
hia_train, hia_valid, hia_test = split_data_hia['train'], split_data_hia['valid'], split_data_hia['test']

Found local copy...
Loading...
Done!
Found local copy...
Loading...
Done!


In [114]:
descriptor_names = list(Chem.rdMolDescriptors.Properties.GetAvailableProperties())
get_descriptors = Chem.rdMolDescriptors.Properties(descriptor_names)
mol = Chem.MolFromSmiles(hia_train.Drug[0])
mol
descriptors = []
if mol:
    descriptors = np.array(get_descriptors.ComputeProperties(mol))

In [117]:
# preprocess the data.
x_hia_train, x_hia_test = drug_descriptors(hia_train.Drug), drug_descriptors(hia_test.Drug)

In [121]:
lda = LinearDiscriminantAnalysis()
lda.fit(x_hia_train, hia_train.Y)

0.9310344827586207


In [ ]:
print("Accuracy classification task:", acc(hia_test.Y, lda.predict(x_hia_test)))

### Regression task freesolv

**Dataset Description**: The Free Solvation Database, FreeSolv(SAMPL), provides experimental and calculated hydration free energy of small molecules in water. The calculated values are derived from alchemical free energy calculations using molecular dynamics simulations. From MoleculeNet.

**Task Description**: Regression. Given a drug SMILES string, predict the activity of hydration free energy. TODO what is hydration free energy?

**Dataset Statistics**: 642 drugs.

**Evaluation Metric**: mean absolute error

In [125]:
data_freesolv = ADME(name = 'hydrationfreeenergy_freesolv')
split_data_freesolv = data_freesolv.get_split(method='scaffold', frac = [0.7, 0.0, 0.3])
freesolv_train, freesolv_valid, freesolv_test = split_data_freesolv['train'], split_data_freesolv['valid'], split_data_freesolv['test']

Found local copy...
Loading...
Done!
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 642/642 [00:00<00:00, 14935.95it/s]


In [126]:
x_freesolv_train, x_freesolv_test = drug_descriptors(freesolv_train.Drug), drug_descriptors(freesolv_test.Drug)

In [146]:
models = [LinearRegression, DecisionTreeRegressor, ExtraTreesRegressor, KNeighborsRegressor, RandomForestRegressor, GradientBoostingRegressor]
for model in models:
    regression_task(model(), x_freesolv_train, freesolv_train.Y, x_freesolv_test, freesolv_test.Y)

85.63802844623862
2.5673125
1.6086259375000005
2.3152125
2.1963840625
2.1567131183678336


In [147]:
model = ExtraTreesRegressor()
regression_task(model, x_freesolv_train, freesolv_train.Y, x_freesolv_test, freesolv_test.Y)

1.6296346874999998


### **Report Results Task A**

| set-data | 
|---|
|validation|
| test |

<a name="section-model1"></a><h2 style="color:rgb(0,120,170)">Task B: Implementation of `Model I` (15 points)</li></a></h2>

<div style="background-color:rgb(224, 243, 255)">

The objective of this task is to implement the first deep learning model, referred to as `Model I`, together with some architectural variations of this model. The task should cover the following points:

**Implementing `Model I` (7 points):** The model can use any of the architectures discussed in the course such as CNNs, RNNs, Attention Networks, Transformers, pretrained Large LMs.

**Model variations (6 points):** Implement **three variations** of `Model I`. Each variation applies only one change to the baseline architecture, making it possible to study the effect of the change. A variation can be a change in for instance the architecture (like number of layers, size of the embeddings, architecture-specific variants, etc.), optimization methods, or regularization approaches. The code of all variations should be inside the code of `Model I`, and executing a variation should be done by simply passing the corresponding parameters of the variation to the model. 


**Reporting and discussion (2 points):** Report the evaluation results of all the variations in a table and also in a plot, accompanied with the result of the dummy baseline. Discuss which variation(s) appear to be the most effective one(s). Explain your take.


</div>

In [ ]:
freesolv_featurizer = MatFeaturizer.from_pretrained('mat_masking_200k')
freesolv_train, freesolv_valid, freesolv_test = get_data_loaders(freesolv_featurizer,
                                          batch_size=32,
                                          task_name='ADME',
                                          dataset_name='hydrationfreeenergy_freesolv')

In [ ]:
# load hia dataset

### **Results discussion**

For both task the regression task turned out to be quite challenging for both the linear model and the transformer. \
Instead, for the classfication task, the trasformer outperforms the linear model significantly by TODO?. 

| Dataset | Metric | Linear model | Transformer
|---|:---| :---:|:---:
|HIV| AUROC | 0.93 | TODO
|FreeSolv| MAE | 1.63 | TODO